### PR TITLE
Refactor CompetenceCourse and ExecutionDegree related methods

### PR DESCRIPTION
**Pull request by pedromonteiro-qubIT**
Relates with: #qubIT-Fenix-8842

### CODE REVIEW

File: src/main/java/org/fenixedu/academic/domain/ExecutionCourse.java [hunk 1]
```diff
@@ -330,15 +315,9 @@ public ExecutionYear getExecutionYear() {
...
-        for (CurricularCourse curricularCourse : getAssociatedCurricularCoursesSet()) {
-            ExecutionDegree executionDegree =
-                    curricularCourse.getDegreeCurricularPlan().getExecutionDegreeByYear(getExecutionYear());
-            if (executionDegree != null) {
-                result.add(executionDegree);
-            }
-        }
-        return result;
+        return getAssociatedCurricularCoursesSet().stream().map(CurricularCourse::getDegreeCurricularPlan)
+                .map(dcp -> dcp.getExecutionDegreeByYear(getExecutionYear())).filter(Objects::nonNull)
```
Comment (eduardofernandes-qubit) [inline src/main/java/org/fenixedu/academic/domain/ExecutionCourse.java]: `getExecutionDegreeByYear` is deprecated. I think we could use `findExecutionDegree` instead

**Review by eduardofernandes-qubit — CHANGES_REQUESTED** [review]

**Review by eduardofernandes-qubit — APPROVED** [review]

**Review by Sofia-Nascimento — APPROVED** [review]

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java [hunk 2]
```diff
@@ -112,4 +124,56 @@ public void getAttendsByStudent_withStudentWithoutRegistrations_returnsNull() {
...
         assertNull(emptyExecutionCourse.getAttendsByStudent(studentWithoutRegistrations));
     }
+
+    @Test
+    public void testGetCompetenceCourses() {
+        final Set<CompetenceCourse> competenceCourses = executionCourse.getCompetenceCourses();
+        assertFalse(competenceCourses.isEmpty());
+        assertEquals(1, competenceCourses.size());
+        final CompetenceCourse competenceCourse = competenceCourses.iterator().next();
+        assertEquals(CompetenceCourseTest.COURSE_A_CODE, competenceCourse.getCode());
```
Comment (sanavarali) [inline src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java:214]: More edge cases should be added:
- execution course with multiple curricular courses of same competence course
- execution course with multiple curricular courses of different competence courses

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java [hunk 3]
```diff
@@ -112,4 +124,56 @@ public void getAttendsByStudent_withStudentWithoutRegistrations_returnsNull() {
...
+    }
+
+    @Test
+    public void testGetCompetenceCoursesInformation() {
```
Comment (sanavarali) [inline src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java:266]: More edge cases should be added:
- execution course with multiple competence courses, and one of them with multiple competence course informations

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java [hunk 4]
```diff
@@ -112,4 +124,56 @@ public void getAttendsByStudent_withStudentWithoutRegistrations_returnsNull() {
...
+        assertEquals(new BigDecimal("6.0"), info.getCredits());
+        assertEquals(AcademicPeriod.SEMESTER, info.getAcademicPeriod());
+
+        final CompetenceCourse cc = executionCourse.getCompetenceCourses().iterator().next();
+        assertTrue(informations.contains(cc.findInformationMostRecentUntil(executionCourse.getExecutionInterval())));
+    }
+
+    @Test
+    public void testGetExecutionDegrees() {
+        final Set<ExecutionDegree> executionDegrees = executionCourse.getExecutionDegrees();
```
Comment (sanavarali) [inline src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java:318]: More edge cases should be added:
- execution course with curricular courses of different degree curricular plans
- and also assert that there are execution degrees created but not related to the execution course

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java [hunk 5]
```diff
@@ -112,4 +124,56 @@ public void getAttendsByStudent_withStudentWithoutRegistrations_returnsNull() {
...
+        assertFalse(executionDegrees.isEmpty());
+        assertEquals(1, executionDegrees.size());
+        assertEquals(Degree.find(DEGREE_A_CODE), executionDegrees.iterator().next().getDegree());
+        assertTrue(executionDegrees.contains(executionDegree));
+    }
+
+    @Test
+    public void testGetAssociatedDegreeCurricularPlans() {
```
Comment (sanavarali) [inline src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java:356]: More edge cases should be added:
- execution course with curricular courses of different degree curricular plans
- and also assert that there are degree curricular plans created but not related to the execution course

**Review by sanavarali — CHANGES_REQUESTED** [review]

**Review by sanavarali — APPROVED** [review]
